File: src/ros2_bridge/mappers/sensor_msgs/battery_state.rs
//! Typed mapper for `sensor_msgs/msg/BatteryState`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn battery_state_to_bus(
    msg: ros_env::sensor_msgs::msg::BatteryState,
) -> crate::sensor_msgs::msg::v1::BatteryState {
    crate::sensor_msgs::msg::v1::BatteryState {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        voltage: msg.voltage,
        current: msg.current,
        charge: msg.charge,
        capacity: msg.capacity,
        design_capacity: msg.design_capacity,
        percentage: msg.percentage,
        power_supply_status: msg.power_supply_status.into(),
        power_supply_health: msg.power_supply_health.into(),
        power_supply_technology: msg.power_supply_technology.into(),
        present: msg.present,
        cell_voltage: crate::ros2_bridge::mappers::convert::f32_seq(msg.cell_voltage),
        cell_temperature: crate::ros2_bridge::mappers::convert::f32_seq(msg.cell_temperature),
        location: crate::ros2_bridge::mappers::convert::from_ros_string(msg.location),
        serial_number: crate::ros2_bridge::mappers::convert::from_ros_string(msg.serial_number),
        temperature: msg.temperature,
    }
}

pub(crate) fn battery_state_to_ros(
    bus: crate::sensor_msgs::msg::v1::BatteryState,
) -> ros_env::sensor_msgs::msg::BatteryState {
    ros_env::sensor_msgs::msg::BatteryState {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(
            bus.header.unwrap_or_default(),
        ),
        voltage: bus.voltage,
        current: bus.current,
        charge: bus.charge,
        capacity: bus.capacity,
        design_capacity: bus.design_capacity,
        percentage: bus.percentage,
        power_supply_status: bus.power_supply_status as _,
        power_supply_health: bus.power_supply_health as _,
        power_supply_technology: bus.power_supply_technology as _,
        present: bus.present,
        cell_voltage: bus.cell_voltage,
        cell_temperature: bus.cell_temperature,
        location: crate::ros2_bridge::mappers::convert::to_ros_string(bus.location),
        serial_number: crate::ros2_bridge::mappers::convert::to_ros_string(bus.serial_number),
        temperature: bus.temperature,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SensorMsgsBatteryStateMapper;

impl TypedTopicMapper for SensorMsgsBatteryStateMapper {
    type Ros = ros_env::sensor_msgs::msg::BatteryState;
    type Bus = crate::sensor_msgs::msg::v1::BatteryState;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(battery_state_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(battery_state_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/sensor_msgs/battery_state.py
"""Generated mapper for `sensor_msgs/msg/BatteryState`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros

def battery_state_to_bus(msg):
    from robot_bus.sensor_msgs.msg.v1 import BatteryState as BusMsg

    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.voltage = msg.voltage
    bus.current = msg.current
    bus.charge = msg.charge
    bus.capacity = msg.capacity
    bus.design_capacity = msg.design_capacity
    bus.percentage = msg.percentage
    bus.power_supply_status = msg.power_supply_status
    bus.power_supply_health = msg.power_supply_health
    bus.power_supply_technology = msg.power_supply_technology
    bus.present = msg.present
    bus.cell_voltage.extend(list(msg.cell_voltage))
    bus.cell_temperature.extend(list(msg.cell_temperature))
    bus.location = str(msg.location)
    bus.serial_number = str(msg.serial_number)
    bus.temperature = msg.temperature
    return bus


def battery_state_to_ros(bus):
    from sensor_msgs.msg import BatteryState as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.voltage = bus.voltage
    out.current = bus.current
    out.charge = bus.charge
    out.capacity = bus.capacity
    out.design_capacity = bus.design_capacity
    out.percentage = bus.percentage
    out.power_supply_status = bus.power_supply_status
    out.power_supply_health = bus.power_supply_health
    out.power_supply_technology = bus.power_supply_technology
    out.present = bus.present
    out.cell_voltage = list(bus.cell_voltage)
    out.cell_temperature = list(bus.cell_temperature)
    out.location = str(bus.location)
    out.serial_number = str(bus.serial_number)
    out.temperature = bus.temperature
    return out


class SensorMsgsBatteryStateMapper:
    def type_name(self) -> str:
        return "sensor_msgs/msg/BatteryState"

    def ros_msg_type(self):
        from sensor_msgs.msg import BatteryState as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return battery_state_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.sensor_msgs.msg.v1 import BatteryState as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return battery_state_to_ros(bus)
